bottlerocket/agents/src/bin/vsphere-k8s-cluster-resource-agent/vsphere_k8s_cluster_provider.rs

```diff
@@ -0,0 +1,763 @@
+
+    /// Name of the VM template for the control plane VMs
+    pub vm_template: String,
+
+    /// The name of the secret containing vCenter credentials.
+    pub vcenter_secret_name: Option<SecretName>,
+}
+
+impl Configuration for ProductionMemo {}
+
+/// Once we have fulfilled the `Create` request, we return information about the cluster we've created
+#[derive(Clone, Debug, Serialize, Deserialize, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct CreatedVSphereK8sCluster {
+    /// The name of the cluster.
+    pub cluster_name: String,
+
+    /// The cluster server endpoint.
+    pub endpoint: String,
+
+    /// Base64 encoded Kubeconfig for the vSphere K8s cluster
+    pub encoded_kubeconfig: String,
+}
+
+impl Configuration for CreatedVSphereK8sCluster {}
+
+pub struct VSphereK8sClusterCreator {}
+
+#[async_trait::async_trait]
+impl Create for VSphereK8sClusterCreator {
+    type Config = VSphereK8sClusterConfig;
+    type Info = ProductionMemo;
+    type Resource = CreatedVSphereK8sCluster;
+
+    async fn create<I>(
+        &self,
+        spec: Spec<Self::Config>,
+        client: &I,
+    ) -> ProviderResult<Self::Resource>
+    where
+        I: InfoClient,
+    {
+        let mut memo: ProductionMemo = client
+            .get_info()
+            .await
+            .context(Resources::Unknown, "Unable to get info from info client")?;
+        // Keep track of the state of resources
+        let mut resources = Resources::Clear;
+
+        // Get vSphere credentials to authenticate to vCenter via govmomi
+        let secret_name = spec
+            .secrets
+            .get(VSPHERE_CREDENTIALS_SECRET_NAME)
+            .context(resources, "Unable to fetch vSphere credentials")?;
+        vsphere_credentials(client, secret_name, &resources).await?;
+        memo.vcenter_secret_name = Some(secret_name.clone());
+
+        // Set current directory to somewhere other than '/' so eksctl-anywhere won't try to mount
+        // it in a container.
+        fs::create_dir_all(WORKING_DIR).context(
+            resources,
+            format!("Failed to create working directory '{}'", WORKING_DIR),
+        )?;
+        env::set_current_dir(Path::new(WORKING_DIR)).context(
+            resources,
+            format!("Failed to change current directory to {}", WORKING_DIR),
+        )?;
+
+        // Check whether cluster creation is necessary
+        let (do_create, message) = is_cluster_creation_required(
+            &spec.configuration.cluster,
+            spec.configuration.creation_policy.unwrap_or_default(),
+        )
+        .await?;
+        memo.current_status = message;
+        info!("{}", memo.current_status);
+        client
+            .send_info(memo.clone())
+            .await
+            .context(resources, "Error sending cluster creation message")?;
+
+        let mgmt_kubeconfig_path = format!("{}/mgmt.kubeconfig", WORKING_DIR);
+        let encoded_kubeconfig = if do_create {
```

Comment:
This is a very large if block. Is this practical? (sometimes this would be impractical if too many arguments and returns are required)

```rust
if do_create {
   create_cluster(x,y,z)?
} else {
  ////
}
```
```

Reply:
I already tried slimming it down with `create_vsphere_k8s_cluster`. I can maybe slim it down further by capturing the preflight/postflight stuff in other functions.